File: libs/lib_base/src/main/java/com/tecsun/robot/nanning/builder/TimeBuilder.java
package com.tecsun.robot.nanning.builder;

import com.tecsun.robot.nanning.lib_base.BaseActivity;
import com.tecsun.robot.nanning.util.log.LogUtil;

import java.util.LinkedList;

public final class TimeBuilder extends BaseBuilder {

    private static final String TAG = TimeBuilder.class.getSimpleName();

    private BaseActivity baseActivity;

    private TimeListener listener;

    private final LinkedList<Integer> list = new LinkedList<>();

    public static final int TIME_1 = 55;

    public static final int TIME_2 = 70;

    public TimeBuilder(BaseActivity baseActivity, TimeListener listener) {
        this.baseActivity = baseActivity;
        this.listener = listener;
        baseActivity.getLifecycle().addObserver(new BuilderLifeCycleObserver(this));
    }


    @Override
    protected void onCreate() {
        mHandler.postDelayed(checkRunnable, 1000);
    }

    @Override
    protected void onStart() {

    }

    @Override
    protected void onResume() {

    }

    @Override
    protected void onPause() {

    }

    @Override
    protected void onDestroy() {
        mHandler.removeCallbacks(checkRunnable);
        super.onDestroy();
    }

    private Runnable checkRunnable = () -> {
        checkTime();
    };

    private final void checkTime() {
        LogUtil.i(TAG, "checkTime()");

        baseActivity.runOnUiThread(() -> {

            list.add(0);

            LogUtil.i(TAG, "list.size() = " + list.size());

            if (listener != null) {
                listener.biz(list.size());
            }

            mHandler.postDelayed(checkRunnable, 1000);
        });

    }

    public interface TimeListener {
        /**
         * 超时多少秒
         *
         * @param second
         */
        void biz(int second);

    }

    public final void initTime() {
        baseActivity.runOnUiThread(() -> {
            LogUtil.i(TAG, "initTime()");
            list.clear();
        });
    }
}

File: libs/lib_base/src/main/java/com/tecsun/robot/nanning/builder/TimeBuilder.kt
package com.tecsun.robot.nanning.builder

import android.os.Handler
import android.os.Looper
import com.tecsun.robot.nanning.util.log.LogUtil
import java.util.*

object TimeBuilder {

    private val TAG = TimeBuilder::class.java.simpleName

    const val TIME_1 = 55

    const val TIME_2 = 70

    private val secondsList = LinkedList<Int>()

    private val mHandler = Handler(Looper.getMainLooper())

    private val bizList: LinkedList<TimeListener?> = LinkedList()


    fun addBiz(l: TimeListener?) {
        LogUtil.i(TAG,"addBiz()")
        mHandler.post {
            secondsList.clear()
            bizList.clear()
            bizList.addFirst(l)
//            mHandler.removeCallbacks(checkRunnable)
        }
    }

    fun startBiz() {
        LogUtil.i(TAG,"startBiz()")
        mHandler.post (checkRunnable)
    }

    private fun doBiz() {
        LogUtil.i(TAG,"doBiz()")
        bizList.first?.biz(secondsList.size)
    }

     fun removeBiz() {
        LogUtil.i(TAG,"removeBiz()")
        mHandler.post {
            bizList.clear()
            secondsList.clear()
            mHandler.removeCallbacks(checkRunnable)
        }
    }

    private val checkRunnable = Runnable {
        checkTime()
    }

    private fun checkTime() {
        LogUtil.i(TAG,">>>>>>>>>>>>>>>>>>>>>>>>>checkTime()")
        secondsList.add(0)
        if (secondsList.size > TIME_2 ) {
            removeBiz()
        } else {
            doBiz()
            mHandler.postDelayed(checkRunnable, 1000)
        }
    }

    interface TimeListener {
        /**
         * 超时多少秒
         * @param second
         */
        fun biz(second: Int)
    }

    fun initTime() {
        LogUtil.i(TAG,"initTime()")
        mHandler.post {
            secondsList.clear()
        }
    }


}